packages/git/src/common/git.ts

```diff
@@ -295,6 +295,28 @@ export namespace Git {
 
         }
 
+        /**
+         * Options for further refining the `git stash` command.
+         */
+        export interface Stash {
+            /**
+             * The kind of stash action.
+             */
+            readonly action?: 'push' | 'apply' | 'pop' | 'list' | 'drop' | 'clear';
+
+            /**
+             * The stash id.
+             * This is an optional argument for actions of kind 'apply', 'pop' and 'drop'.
```

Comment:
Wouldn't it make sense to have the `id` only when `action` is one of `'apply', 'pop' and 'drop'`? Same for `message`. If it is not applicable for an `action` do not make it available.

You could alter the return type based on the arguments similarly we did for `branch`, `remote` and `lsFiles`.

Reply:
Let's leave this as is and split up the methods. Would this work?

```ts
/**
 * The default `git stash` command. Equivalent to `git stash push`. If the `message` is not defined, the Git default *WIP on branchname* will be used instead.
 */
stash(repository: Repository, options?: Readonly<{ action?: 'push', message?: string }>): Promise<void>;

/**
 * Resolves to an array of stashed entries that you currently have. Same as `git stash list`.
 */
stash(repository: Repository, options: Readonly<{ action: 'list' }>): Promise<Git.StashEntry[]>;

/**
 * Removes all the stash entries.
 */
stash(repository: Repository, options: Readonly<{ action: 'clear' }>): Promise<void>;

/**
 * Like `git stash pop`, but does not remove the state from the stash list.
 */
stash(repository: Repository, options: Readonly<{ action: 'apply', id?: string }>): Promise<void>;

/**
 * Removes a single stashed state from the stash list and applies it on top of the current working tree state.
 * The single stashed state is identified by the optional `id`. If the `id` is not defined the latest stash will be popped.
 */
stash(repository: Repository, options: Readonly<{ action: 'pop', id?: string }>): Promise<void>;

/**
 * Removes a single stash entry from the list of stash entries. When the `id` is not given, it removes the latest one.
 */
stash(repository: Repository, options: Readonly<{ action: 'drop', id?: string }>): Promise<void>;
```

Reply:
We can still split up the stash options if required but I would rather drop it then. Let me know if the proposed API is inconvenient on the call-sites.

Reply:
Ok, I changed as you proposed, with one exception: I had to merge
```
stash(repository: Repository, options: Readonly<{ action: 'apply', id?: string }>): Promise<void>;
stash(repository: Repository, options: Readonly<{ action: 'pop', id?: string }>): Promise<void>;
stash(repository: Repository, options: Readonly<{ action: 'drop', id?: string }>): Promise<void>;
```
to
```
stash(repository: Repository, options: Readonly<{ action: 'apply' | 'pop' | 'drop', id?: string }>): Promise<void>;
```
to make it work with code like [here](https://github.com/theia-ide/theia/blob/14f7ed3c890b26007aa072d9c63183fc337de230/packages/git/src/browser/git-quick-open-service.ts#L374).
All in all the code assistance is not very proper this way. If I want to know what types of actions are possible to use I just get `push` offered even if the other actions work. At least the compiler says 'NO' if you try to use it the wrong way.